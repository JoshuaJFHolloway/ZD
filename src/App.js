import React, { Component } from 'react';
import './App.css';

class App extends Component {
  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Ned Ned's Cheese Emporium</h1>
        </header>
        <span role={"img"} aria-label={"emoji"} >Today you can buy 12 blocks of cheese for Â£19.99 ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€</span>
      </div>
    );
  }
}

export default App;
