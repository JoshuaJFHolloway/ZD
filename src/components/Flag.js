import React from 'react';

const Flag = (props) => {
  return (
    <div>
      <img
        src={props.image}
        width="500px"
        onClick={props.onClick}
      />
    </div>
  )
};

export default Flag;