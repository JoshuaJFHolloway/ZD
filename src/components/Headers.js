import React from 'react';
import Header from './Header';
import HeaderStyle from './styledComponents/Header.js'
import SubHeaderStyle from './styledComponents/subHeader.js'

const Headers = () => {
  return (
    <div>
      <HeaderStyle>
        <Header
          title={"Ned Ned's Cheese Emporium"}
        />
      </HeaderStyle>
      <SubHeaderStyle>
        <Header role={"img"} aria-label={"emoji"}
          title={"Today you can buy 12 blocks of cheese for Â£19.99 ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€ğŸ§€"}
        />
      </SubHeaderStyle>
    </div>
  )
};

export default Headers;