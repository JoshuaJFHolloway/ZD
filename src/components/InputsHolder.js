import React from 'react';
import Header from './Header';
import InputTitleStyle from './styledComponents/inputTitle';
import ButtonStyle from './styledComponents/Button';
import InputStyle from './styledComponents/Input';

const InputsHolder = (props) => {
  return (
    <div>
      <InputTitleStyle>
        <Header
          title={props.title}
        />
      </InputTitleStyle>

      <InputStyle
        type={props.type}
        name={props.inputName}
        onChange={props.handleChange}
      />

      <ButtonStyle
        name={props.buttonName}
        onClick={props.handleClick}
      >
        {props.buttonName}
      </ButtonStyle>
    </div>
  )
};

export default InputsHolder;