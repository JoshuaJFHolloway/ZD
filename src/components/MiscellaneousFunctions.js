export function currencyConverter(rates) {
  const { GBP, EUR } = rates;
  let convertedCurrency = null;

  if (rates) {
    console.log("Rates:", rates);
    console.log("EUR", EUR);
    console.log("GBP", GBP);

    convertedCurrency = EUR / GBP;
  }

  return convertedCurrency;
}

export function cheeseBlockCalculator(amount, rates) {
  const euros = currencyConverter(rates) * amount;
  const cheeseBlock = 3.24;

  return euros / cheeseBlock;
}

