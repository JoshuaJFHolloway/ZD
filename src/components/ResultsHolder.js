import React from 'react';
import ResultsStyle from './styledComponents/Results';

const ResultsHolder = (props) => {

  const { state } = props;

  if (!state || Object.keys(state.results).length === 0) {
    return <ResultsStyle>The number of blocks of cheese you can afford will appear here...</ResultsStyle>
  }

    const currencyConverter = () => {
      const rates = state.results;
      const GBP = rates.GBP;
      const EUR = rates.EUR;

      return EUR / GBP;
    };

    const cheeseBlockCalculator = () => {
      const cheeseMoney = state.params.cheeseMoney;
      const euros = currencyConverter() * cheeseMoney;
      const cheeseBlock = 3.24;

      return euros / cheeseBlock;
    };

    const cheeseBlocks = cheeseBlockCalculator();

    return (
      <div>
        <ResultsStyle>
          {Math.floor(cheeseBlocks)} blocks of cheese!
        </ResultsStyle>
      </div>
    )
};

export default ResultsHolder;