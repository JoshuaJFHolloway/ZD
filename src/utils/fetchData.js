// would include 'request' as one of the three arguments if needed to pass param into API URL

const fetchData = (onSuccess, onFailure) => {
  const API_URL = "https://openexchangerates.org/api/latest.json?app_id=66509a8c337e4718b4f3c4c235c5db69";
  const apiConfig = {
    method: 'get',
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    },
    // body: request
  };

  fetch(API_URL, apiConfig)
    .then(results => results.json())
    .then(response => {
      onSuccess && onSuccess(response);
    })
    .catch(error => {
      onFailure && onFailure(error);
    });
};

export {fetchData}