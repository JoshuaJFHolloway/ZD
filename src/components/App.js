import React, { Component } from 'react';
import Headers from './Headers';
import Inputs from './Inputs';
import Results from './Results';

class App extends Component {
  render() {
    return (
      <div>
        <Headers/>
        <Inputs/>
        <Results/>
      </div>
    );
  }
}

export default App;
