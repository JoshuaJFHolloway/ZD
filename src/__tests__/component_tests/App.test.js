import React from 'react';
import {shallow} from 'enzyme';
import App from './../../components/App';
import sinon from 'sinon';

import * as MiscellaneousFunctions from '../../components/MiscellaneousFunctions';
import * as fetchUtils from '../../utils/fetchData';
import {rates} from '../../utils/mockData/response';

let app;

const handleChangeEvent = {
  target: {
    name: 'cheeseMoney',
    value: '20'
  }
};

const GBEvent = {
  target: {
    name: "GB",
    value: true
  }
};

const JapanEvent = {
  target: {
    name: "Japan",
    value: true
  }
};


const response = {
  rates: rates
};

describe('App', () => {

  let fetchStub;
  let onSuccessStub;
  let onFailureStub;
  let areParamsValidStub;
  let cheeseBlockCalculatorStub;

  beforeAll(() => {
    fetchStub = sinon.stub(fetchUtils, "fetchData");
  });

  beforeEach(() => {
    app = shallow(<App/>);

    onSuccessStub = sinon.stub(app.instance(), "onSuccess");
    onFailureStub = sinon.stub(app.instance(), "onFailure");
  });

  afterEach(() => {
    fetchStub.reset();
  });

  afterAll(() => {
    fetchStub.restore();
  });

  it('renders correctly', () => {
    expect(app).toMatchSnapshot();
  });

  describe('rendering components', () => {

      it('has headers', () => {
        expect(app.find('Headers'));
      });

      it('has InputsHolder', () => {
        expect(app.find('InputsHolder'));
      });

      it('has ResultsHolder', () => {
        expect(app.find('ResultsHolder'));
      });

      it('has two flag components', () => {
        expect(app.find('Flag')).at(0).prop('name').toBe("Japan");
        expect(app.find('Flag')).at(1).prop('name').toBe("GB");
      })
  });

  describe('Initializes state', () => {

    it("initializes cheeseMoney state as '' ", () => {
      expect(app.state('cheeseMoney')).toEqual('');
    });

    it('initializes amountOfCheese state as 0', () => {
      expect(app.state('amountOfCheese')).toEqual(0);
    });

    it('initializes rates state as undefined', () => {
      expect(app.state('rates')).toEqual(undefined);
    });

    it('initializes GB state as false', () => {
      expect(app.state('GB')).toEqual(false);
    });

    it('initializes Japan state as false', () => {
      expect(app.state('Japan')).toEqual(false);
    });
  });

  describe('Updating state successfully', () => {

    describe('handleChange', () => {

      it('updates the state correctly depending on the event.target', () => {
        app.instance().handleChange(handleChangeEvent);
        expect(app.instance().state.cheeseMoney).toEqual('20');
      });

      it('updates the state of GB to true if clicked', () => {
        app.instance().handleChange(GBEvent);
        expect(app.instance().state.GB).toEqual(true)
      });

      it('updates the state of Japan to true if clicked', () => {
        app.instance().handleChange(JapanEvent);
        expect(app.instance().state.Japan).toEqual(true)
      });
    });
  });

  describe('areParamsValid', () => {

    it('should return false if cheeseMoney has no length', () => {
      app.instance().state.cheeseMoney = '';
      expect(app.instance().areParamsValid()).toEqual(false);
    });

    it('should return true if cheeseMoney has a value', () => {
      app.instance().state.cheeseMoney = '30';
      expect(app.instance().areParamsValid()).toEqual(true);
    })
  });

  describe('Getting data', () => {
    it('should call the fetch stub once', () => {
      areParamsValidStub = sinon.stub(app.instance(), "areParamsValid");
      areParamsValidStub.returns(true);
      app.instance().getData();
      expect(fetchStub.calledOnce).toBe(true);
    });
  });

  describe('onSuccess', () => {
    it('should update the state with the rates and amount of cheese', () => {
      cheeseBlockCalculatorStub = sinon.stub(MiscellaneousFunctions, 'cheeseBlockCalculator');
      cheeseBlockCalculatorStub.returns(10);
      app.instance().onSuccess(response);
      expect(cheeseBlockCalculatorStub.calledOnce).toBe(true);
      expect(app.instance().state.amountOfCheese).toEqual(10);
    })
  });

});
