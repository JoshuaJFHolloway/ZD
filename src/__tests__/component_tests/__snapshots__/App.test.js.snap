// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Headers />,
        <div>
          <styled.h3>
                    <InputsHolder
                              buttonName="Convert GBP to cheese"
                              handleChange={[Function]}
                              handleClick={[Function]}
                              inputName="cheeseMoney"
                              title="Enter in your amount in GBP"
                              type="number"
                    />
          </styled.h3>
          <styled.h3>
                    <ResultsHolder
                              state={
                                        Object {
                                                  "params": Object {
                                                    "buttonClicked": false,
                                                    "cheeseMoney": false,
                                                  },
                                                  "results": Object {},
                                                }
                              }
                    />
          </styled.h3>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <styled.h3>
              <InputsHolder
                            buttonName="Convert GBP to cheese"
                            handleChange={[Function]}
                            handleClick={[Function]}
                            inputName="cheeseMoney"
                            title="Enter in your amount in GBP"
                            type="number"
              />
</styled.h3>,
            <styled.h3>
              <ResultsHolder
                            state={
                                          Object {
                                                        "params": Object {
                                                          "buttonClicked": false,
                                                          "cheeseMoney": false,
                                                        },
                                                        "results": Object {},
                                                      }
                            }
              />
</styled.h3>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <InputsHolder
                buttonName="Convert GBP to cheese"
                handleChange={[Function]}
                handleClick={[Function]}
                inputName="cheeseMoney"
                title="Enter in your amount in GBP"
                type="number"
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "buttonName": "Convert GBP to cheese",
                "handleChange": [Function],
                "handleClick": [Function],
                "inputName": "cheeseMoney",
                "title": "Enter in your amount in GBP",
                "type": "number",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <ResultsHolder
                state={
                                Object {
                                                "params": Object {
                                                  "buttonClicked": false,
                                                  "cheeseMoney": false,
                                                },
                                                "results": Object {},
                                              }
                }
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "state": Object {
                  "params": Object {
                    "buttonClicked": false,
                    "cheeseMoney": false,
                  },
                  "results": Object {},
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Headers />,
          <div>
            <styled.h3>
                        <InputsHolder
                                    buttonName="Convert GBP to cheese"
                                    handleChange={[Function]}
                                    handleClick={[Function]}
                                    inputName="cheeseMoney"
                                    title="Enter in your amount in GBP"
                                    type="number"
                        />
            </styled.h3>
            <styled.h3>
                        <ResultsHolder
                                    state={
                                                Object {
                                                            "params": Object {
                                                              "buttonClicked": false,
                                                              "cheeseMoney": false,
                                                            },
                                                            "results": Object {},
                                                          }
                                    }
                        />
            </styled.h3>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <styled.h3>
                <InputsHolder
                                buttonName="Convert GBP to cheese"
                                handleChange={[Function]}
                                handleClick={[Function]}
                                inputName="cheeseMoney"
                                title="Enter in your amount in GBP"
                                type="number"
                />
</styled.h3>,
              <styled.h3>
                <ResultsHolder
                                state={
                                                Object {
                                                                "params": Object {
                                                                  "buttonClicked": false,
                                                                  "cheeseMoney": false,
                                                                },
                                                                "results": Object {},
                                                              }
                                }
                />
</styled.h3>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <InputsHolder
                  buttonName="Convert GBP to cheese"
                  handleChange={[Function]}
                  handleClick={[Function]}
                  inputName="cheeseMoney"
                  title="Enter in your amount in GBP"
                  type="number"
/>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "buttonName": "Convert GBP to cheese",
                  "handleChange": [Function],
                  "handleClick": [Function],
                  "inputName": "cheeseMoney",
                  "title": "Enter in your amount in GBP",
                  "type": "number",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <ResultsHolder
                  state={
                                    Object {
                                                      "params": Object {
                                                        "buttonClicked": false,
                                                        "cheeseMoney": false,
                                                      },
                                                      "results": Object {},
                                                    }
                  }
/>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "state": Object {
                    "params": Object {
                      "buttonClicked": false,
                      "cheeseMoney": false,
                    },
                    "results": Object {},
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
